test/envtest/admission_webhook_envtest_test.go

```diff
@@ -0,0 +1,213 @@
+		},
+		{
+			name: "should fail the validation if the spec.config does not pass the schema check of Kong gateway",
+			kongVault: &kongv1alpha1.KongVault{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "vault-invalid-config",
+					Annotations: map[string]string{
+						annotations.IngressClassKey: annotations.DefaultIngressClass,
+					},
+				},
+				Spec: kongv1alpha1.KongVaultSpec{
+					Backend:     "env",
+					Prefix:      "invalid-config",
+					Description: "test env vault",
+					Config: apiextensionsv1.JSON{
+						Raw: []byte(`{"prefix":"kong-env-test","foo":"bar"}`),
+					},
+				},
+			},
+			expectErrorContains: `vault configuration in invalid: schema violation (config.foo: unknown field)`,
+		},
+		{
+			name: "should fail the validation if spec.prefix is duplicate",
+			kongVault: &kongv1alpha1.KongVault{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "vault-dupe",
+					Annotations: map[string]string{
+						annotations.IngressClassKey: annotations.DefaultIngressClass,
+					},
+				},
+				Spec: kongv1alpha1.KongVaultSpec{
+					Backend:     "env",
+					Prefix:      prefixForDuplicationTest, // This is the same prefix as the one created in setup.
+					Description: "test env vault",
+				},
+			},
+			expectErrorContains: fmt.Sprintf(`spec.prefix "%s" is duplicate with existing KongVault`, prefixForDuplicationTest),
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := ctrlClient.Create(ctx, tc.kongVault)
+			if tc.expectErrorContains != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), tc.expectErrorContains)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+// prepareKongVaultAlreadyProgrammedInGateway creates a KongVault and waits until it gets programmed in Gateway.
+func prepareKongVaultAlreadyProgrammedInGateway(
+	ctx context.Context,
+	t *testing.T,
+	ctrlClient client.Client,
+	vaultPrefix string,
+) {
+	t.Helper()
+
+	name := uuid.NewString()
+	vault := &kongv1alpha1.KongVault{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Annotations: map[string]string{
+				annotations.IngressClassKey: annotations.DefaultIngressClass,
+			},
+		},
+		Spec: kongv1alpha1.KongVaultSpec{
+			Backend:     "env",
+			Prefix:      vaultPrefix,
+			Description: "vault description",
+		},
+	}
+	err := ctrlClient.Create(ctx, vault)
+	require.NoError(t, err)
+
+	t.Logf("Waiting for KongVault %s to be programmed...", name)
+	require.Eventuallyf(t, func() bool {
+		kv := &kongv1alpha1.KongVault{}
+		err := ctrlClient.Get(ctx, k8stypes.NamespacedName{Name: name}, kv)
+		if err != nil {
+			return false
+		}
+		programmed, ok := lo.Find(kv.Status.Conditions, func(c metav1.Condition) bool {
+			return c.Type == "Programmed"
+		})
+		return ok && programmed.Status == metav1.ConditionTrue
+	}, 5*time.Second, time.Millisecond*20, "KongVault %s was expected to be programmed", name)
```

Comment:
Won't 5s wait time end up being a bit flake? That is: isn't the time a bit too short? Aren't we afraid that this will flake?

Reply:
From my runs, I didn't experience it being not enough, but well, it won't hurt if we bump it to `30s`. :) https://github.com/Kong/kubernetes-ingress-controller/pull/5605/commits/4a88ffc9295bcb83d60f14bc9ac8f7dc1474e888